### PR TITLE
Refactor the houston Package

### PR DESCRIPTION
**Pull request by AmFlint**
## Description

Refactor the `houston` package to be a high-level interface to interact with the Houston GraphQL API.

Clean the code, and make it more maintainable.

## 🎟 Issue(s)

https://github.com/astronomer/issues/issues/4161

## 🧪 Functional Testing

> List the functional testing steps to confirm this feature or fix.

## 📸 Screenshots

> Add screenshots to illustrate the validity of these changes.

## 📋 Checklist

- [x] Rebased from the main (or release if patching) branch (before testing)
- [x] Ran `make test` before taking out of draft
- [x] Added/updated applicable tests
- [x] Tested against [Houston-API](https://github.com/astronomer/houston-api/) and [Astronomer](https://github.com/astronomer/astronomer/) (if necessary).
- [ ] Communicated to/tagged owners of respective clients potentially impacted by these changes.
- [ ] Updated any related [documentation](https://github.com/astronomer/docs/)

### CODE REVIEW

File: auth/auth.go [hunk 1]
```diff
@@ -17,22 +17,12 @@ import (
 )
 
 // basicAuth handles authentication with the houston api
-func basicAuth(username, password string) (string, error) {
+func basicAuth(username, password string, client houston.HoustonClientInterface) (string, error) {
```
Comment (andriisoldatenko) [inline auth/auth.go:20]:
```suggestion
func basicAuth(username, password string, client houston.ClientInterface) (string, error) {
```

wdyt to rename to ClientInterface? since it's already in houston package.

File: houston/auth.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	acReq := Request{
+		Query: AuthConfigGetRequest,
+	}
+	
+	acResp, err := acReq.DoWithClient(h.client)
+	if err != nil {
+		return nil, err
+	}
+	return acResp.Data.GetAuthConfig, nil
+}
```
Comment (andriisoldatenko) [inline houston/auth.go:29]:
```suggestion
}

```

File: houston/houston.go [hunk 3]
```diff
@@ -17,12 +17,73 @@ var (
...
+	CreateServiceAccountInWorkspace(variables CreateServiceAccountRequest) (*WorkspaceServiceAccount, error)
+	DeleteServiceAccountFromWorkspace(workspaceID, serviceAccountID string) (*ServiceAccount, error)
+	ListServiceAccountsInWorkspace(workspaceID string) ([]ServiceAccount, error)
+	// app
+	GetAppConfig() (*AppConfig, error)
+	GetAvailableNamespaces() ([]Namespace, error)
+}
+
+// HoustonClientImplementation - implementation of the Houston Client Interface
+type HoustonClientImplementation struct {
```
Comment (andriisoldatenko) [inline houston/houston.go:67]: houston.Houston a bit wordy IMO.

File: houston/houston.go [hunk 4]
```diff
@@ -17,12 +17,73 @@ var (
 	ErrVerboseInaptPermissions = errors.New("you do not have the appropriate permissions for that: Your token has expired. Please log in again")
 )
 
+// HoustonClientInterface - Interface that defines methods exposed by the Houston API
+type HoustonClientInterface interface {
```
Comment (andriisoldatenko) [inline houston/houston.go:21]:
```suggestion
type ClientInterface interface {
```

File: houston/houston.go (hunk 3 above)
Comment (andriisoldatenko) [inline houston/houston.go:67]:
```suggestion
type ClientImplementation struct {
```

File: houston/serviceaccount.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
+package houston
```
Comment (andriisoldatenko) [inline houston/serviceaccount.go:1]: rename file to service_account.go pls

File: houston/serviceaccount.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+	resp, err := req.DoWithClient(h.client)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp.Data.CreateDeploymentServiceAccount, nil
+}
+
+// CreateServiceAccountInWorkspace - create a service account into a workspace
+func (h HoustonClientImplementation) CreateServiceAccountInWorkspace(variables CreateServiceAccountRequest) (*WorkspaceServiceAccount, error) {
```
Comment (andriisoldatenko) [inline houston/serviceaccount.go:27]:
```suggestion
func (h HoustonClientImplementation) CreateWorkspaceServiceAccount(variables CreateServiceAccountRequest) (*WorkspaceServiceAccount, error) {
```

File: houston/serviceaccount.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	resp, err := req.DoWithClient(h.client)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp.Data.CreateWorkspaceServiceAccount, nil
+}
+
+// DeleteServiceAccountFromDeployment - delete a service account from a workspace
+func (h HoustonClientImplementation) DeleteServiceAccountFromDeployment(deploymentID, serviceAccountID string) (*ServiceAccount, error) {
```
Comment (andriisoldatenko) [inline houston/serviceaccount.go:41]:
```suggestion
func (h HoustonClientImplementation) DeleteDeploymentServiceAccount(deploymentID, serviceAccountID string) (*ServiceAccount, error) {
```

**Comment by andriisoldatenko** [comment]
@AmFlint pls resolve conflicts.

File: houston/houston.go [hunk 8]
```diff
@@ -18,12 +18,73 @@ var (
...
+}
+
+// ClientImplementation - implementation of the Houston Client Interface
+type ClientImplementation struct {
+	client *Client
+}
+
+// Init - initialized the Houston Client singleton object with proper HTTP Client configuration
+// set as a variable so we can change it to return mock houston clients
+// TODO: RENAME THIS METHOD ONCE TESTS ARE REFACTORED TOO
```
[the remark this answers is not in the document]
Comment (AmFlint) [reply houston/houston.go:74]: Sorry about that, I renamed the `Init` method to `NewClient`, and the old `NewHoustonClient` method to `newInternalClient` (as this becomes a structure internal to the houston package).

File: Makefile [hunk 9]
```diff
@@ -59,6 +59,10 @@ uninstall:
 	$(eval DESTDIR ?= $(GOBIN))
 	rm $(GOBIN)/$(OUTPUT)
 
+# TODO: WE MIGHT WANT TO ADD A TOP-LEVEL mock COMMAND TO GENERATE ALL OF THE MOCK INTERFACES
+mock_houston:
+	mockery --filename=ClientInterface.go --output=airflow/mocks --dir=houston --name ClientInterface
```
Comment (neel-astro) [inline Makefile:64]: sorry I didn't realise this earlier, but it would be better to keep mocks of an interface either where interface is defined or where it is being used, so you might want to move this mocks to `houston` folder instead of `airflow`

**Review by neel-astro — APPROVED** [review]
lgtm 👍 , have added a minor comment regarding placing mocks, also do have a look if there are any easy ways to bring up the dropping code coverage in this PR.

**Comment by AmFlint** [comment]
The code coverage drops in this PR because I've added more files.

I added and refactored tests in the other issue.

However, I don't see a value in testing every Houston method since all it does is call doWithClient and return a structure from the response

File: Makefile (hunk 9 above)
Comment (AmFlint) [reply Makefile:64]: Yeah I agree with you, I just wanted to keep the same conventions already configured in the project.

**Comment by andriisoldatenko** [comment]
> The code coverage drops in this PR because I've added more files.
> 
> I added and refactored tests in the other issue.
> 
> However, I don't see a value in testing every Houston method since all it does is call doWithClient and return a structure from the response

i don't understand you point, i think we want keep coverage as is. Please add tests to fix code coverage at least.

**Comment by AmFlint** [comment]
> > The code coverage drops in this PR because I've added more files.
> > I added and refactored tests in the other issue.
> > However, I don't see a value in testing every Houston method since all it does is call doWithClient and return a structure from the response
> 
> i don't understand you point, i think we want keep coverage as is. Please add tests to fix code coverage at least.

I added tests for the houston package
